File: python/atelier_core/support/schedule.py
"""
Technician schedule: a `@property` with a setter, a private set, and the
methods `next_slot` / `book_slot` that Technician forwards to through
`__getattr__` (see concerns/forwards_to_schedule.py).
"""

from __future__ import annotations


class Schedule:
    DEFAULT_CAPACITY = 16

    def __init__(self, capacity: int = DEFAULT_CAPACITY) -> None:
        self._capacity = capacity
        self._booked: set[int] = set()

    @property
    def capacity(self) -> int:
        return self._capacity

    @capacity.setter
    def capacity(self, value: int) -> None:
        self._capacity = max(value, len(self._booked))

    def next_slot(self) -> int | None:
        """First free slot index, or None when fully booked."""
        for slot in range(self._capacity):
            if slot not in self._booked:
                return slot
        return None

    def book_slot(self, slot: int) -> bool:
        if slot >= self._capacity or slot in self._booked:
            return False
        self._booked.add(slot)
        return True

    def booked_count(self) -> int:
        return len(self._booked)

File: javascript/packages/core/src/support/schedule.js
/**
 * Technician scheduling.
 *
 * A technician forwards to this class through a Proxy, so
 * `technician.nextSlot()` has no declaration on the technician side. See
 * `../concerns/forwardsToSchedule.js`.
 */

/**
 * Half-hour slot index inside a working day (0 == 08:00).
 *
 * @typedef {number} SlotIndex
 */

export class Schedule {
    /** Slots in a working day when no explicit capacity is given. */
    static DEFAULT_CAPACITY = 16;

    /** @type {Set<SlotIndex>} */
    #booked = new Set();

    /** @type {number} */
    #capacity;

    /** @param {number} [capacity] */
    constructor(capacity = Schedule.DEFAULT_CAPACITY) {
        this.#capacity = capacity;
    }

    /**
     * Slots this day can hold.
     *
     * @returns {number}
     */
    get capacity() {
        return this.#capacity;
    }

    /**
     * Shrinking the day never drops an already booked slot.
     *
     * @param {number} value
     */
    set capacity(value) {
        this.#capacity = Math.max(value, this.#booked.size);
    }

    /**
     * First free slot, or undefined when the day is full.
     *
     * @returns {SlotIndex | undefined}
     */
    nextSlot() {
        for (let slot = 0; slot < this.#capacity; slot += 1) {
            if (!this.#booked.has(slot)) {
                return slot;
            }
        }

        return undefined;
    }

    /**
     * Book `slot`; returns false when it was already taken or out of range.
     *
     * @param {SlotIndex} slot
     * @returns {boolean}
     */
    bookSlot(slot) {
        if (slot >= this.#capacity || this.#booked.has(slot)) {
            return false;
        }

        this.#booked.add(slot);

        return true;
    }

    /** @returns {number} */
    bookedCount() {
        return this.#booked.size;
    }
}
